ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/RecoverySuspendTest.java

```diff
@@ -0,0 +1,199 @@
+            _manager.resume();
+        }
+
+        // Let's clean
+        ThreadActionData.purgeActions();
+
+        if (!TxControl.isEnabled()) {
+            // Re-activate transaction creation
+            TxControl.enable();
+        }
+    }
+
+    @Test
+    @BMScripts(scripts = {
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_BytemanControlledXAResource"),
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_PeriodicRecovery")
+    })
+    public void testSuspensionWhenThereArentTransactionsToRecover() {
+
+        // Make sure that the test environment is ready
+        BytemanControlledXAResource.resetCommitCallCounter();
+        BytemanControlledXAResource.resetRollbackCallCounter();
+        // The following instruction makes the BytemanControlledXAResource commit without failures.
+        BytemanControlledXAResource.setGreenFlag();
+        BytemanControlledXAResource.setCommitReturn(XAResource.XA_OK);
+
+        // BytemanControlledXAResource.getCommitCallCounter() should be 1 as:
+        // - One invocation from the normal commit procedure should pass
+        startTest(1,0, 0);
+    }
+
+    @Test
+    @BMScripts(scripts = {
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_BytemanControlledXAResource"),
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_FailTest")
+    })
+    public void testSuspensionWhenThereIsXATxnToRecover() {
+
+        // Make sure that the test environment is ready
+        BytemanControlledXAResource.resetCommitCallCounter();
+        BytemanControlledXAResource.resetRollbackCallCounter();
+        BytemanControlledXAResource.resetGreenFlag();
+        BytemanControlledXAResource.setCommitReturn(XAException.XAER_RMFAIL);
+
+        // BytemanControlledRecord.getCommitCallCounter() should be 3 as:
+        // - One invocation from the normal commit procedure should fail
+        // - One invocation from the recovery process should fail
+        // - One invocation from the recovery process should pass
+        startTest(1, 2, 0);
+    }
+
+    private void startTest(int numberOfCommits, int numberOfRollbacks, int timeout) {
+
+        // obtain a new RecoveryManager with the above config:
+        _manager = RecoveryManager.manager(RecoveryManager.INDIRECT_MANAGEMENT);
+
+        createXATransaction(new BytemanControlledXAResource(), timeout);
+
+        // The Transaction System needs to be disabled, i.e. no new transactions can be created
+        TxControl.disable();
+
+        // Makes sure that the transaction reaper completed all transactions
+        TransactionReaper.transactionReaper().waitForAllTxnsToTerminate();
+
+        // Makes sure that all transactions without a timeout have passed the `prepare` phase
+        waitForTransactionsWithoutTimeout();
+
+        // Synchronization is not needed (i.e. async = true) as isWaitForWorkLeftToDo() == true
+        _manager.suspend(true);
```

Comment:
Please can I ask you to draw my attention how we know that the rollback/commit we might look for occurred during the suspend call?

Reply:
Orphan transactions will be processed only after an interval that tells XARecoveryModule that it can use the collection of OrphanFilters to go through left-over XIDs. Creating/Starting the Recovery Manager just before an XA transaction gets created ensures that the XARecoveryModule will invoke `rollback` after `manager.suspend(true)`

Reply:
IIUC then you are relying on the timing of the recovery manager thread running? I wonder if we could use byteman to be sure but I understand the assumption.

That being said - the way these intervals are set up I think it's quite close that the recovery cycle could take place before this call. From what I can see a small delay of perhaps (guessing what to add together) 6 (so the config of 5 and the config of 1 both higher in this file) seconds might be enough to not make this test do what we expect and possibly recover before the suspend starts

I would prefer byteman to be sure of what is happening. But perhaps a more appropriate configurable assumption to use would be https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/common/RecoveryEnvironmentBean.java#L131 - but like I say, that is still an assumption. Really it would be good to see the recovery manager suspend and the rollback call being coordinated by byteman.

Reply:
I also thought to use byteman to synchronise when the first recovery scan should take place but I have the feeling that, at the end, we don't really need that level of control. The reason why I think so is that the XID returned by the `BytemanControlledXAResource` during recovery will be considered by the XARecoveryModule only after `periodicRecoveryPeriod * 2`, i.e. 10secs. Thus, even if there are multiple scans in the time window between the creation of the Recovery Manager and the invocation of `manager.suspend()`, the XARecoveryModule will process the leftover XID only during the Recovery Manager's suspension.

Reply:
At least, that is my understanding. Please, let me know if I got it wrong :-)

Reply:
So if I understand correctly you are saying that because the recovery will take (say) 10 seconds minimally you expect to reach this instruction before that point and thus it should be considered safe?

Personally I think 10 seconds is not much different to 6 seconds but I suppose it's a bit bigger. I would worry a bit that on CI createXATransaction and waitForAllTxnsToTerminate and waitForTransactionsWithoutTimeout could combine to take longer than 10 seconds.

I don't really like not using Byteman for this and just relying on timeouts but maybe you could argue a 10 second timeout would (generally) be long enough if the instruction `_manager = RecoveryManager.manager(RecoveryManager.INDIRECT_MANAGEMENT);` were to be moved immediately prior to the `_manager.suspend(true);`

Reply:
> So if I understand correctly you are saying that because the recovery will take (say) 10 seconds minimally you expect to reach this instruction before that point and thus it should be considered safe?
>
> Personally I think 10 seconds is not much different to 6 seconds but I suppose it's a bit bigger. I would worry a bit that on CI createXATransaction and waitForAllTxnsToTerminate and waitForTransactionsWithoutTimeout could combine to take longer than 10 seconds.

In a production environment, I would agree with you 100pc. In this test class, createXATransaction, waitForAllTxnsToTerminate and waitForTransactionsWithoutTimeout will take much less than a second as there is only one transaction (without timeout) and it is immediately committed (as the test uses local dummy XAResources).

> I don't really like not using Byteman for this and just relying on timeouts but maybe you could argue a 10 second timeout would (generally) be long enough if the instruction `_manager = RecoveryManager.manager(RecoveryManager.INDIRECT_MANAGEMENT);` were to be moved immediately prior to the `_manager.suspend(true);`?

I was thinking about this while I was reading your comment :-) Although I'm a Byteman fanatic, I have the feeling that using a rendezvous in this case is an overkill. Nevertheless, if you have a strong preference for Byteman, I will employ it :-)

Reply:
> I kind of think that on CI it is possible for delays in processing to happen so I feel that either moving the start of the recovery manager to just before here or doing Byteman should be done. I am stopping short of saying either are mandatory.

`_manager = RecoveryManager.manager(RecoveryManager.INDIRECT_MANAGEMENT)` moved before `_manager.suspend()`. Thanks Tom

> The main part of the pull request that I think we do need to address is the thread where I talk about testSuspensionWhenThereIsAnXAResourceToCommit - being confident the server waits for transactions (that have an XAR) to commit is definitely an important part to prove in my opinion).

About `testSuspensionWhenThereIsXATxnToRecover`, Byteman makes sure that we pass two checkpoints before the test can succeed:
1. `RULE setwaitForWorkLeftToDoFlag` sets the flag `waitForWorkLeftToDoFlag` only when `PeriodicRecovery.suspendScan` was invoked while `_waitForWorkLeftToDo` was set to true
2. `RULE RULE setGreenFlag` sets the flag `greenFlag` only when point 1. is verified and `XARecoveryModule.hasWorkLeftToDo()` returns true (i.e. there is at least one transaction to recover)
3. `RULE FailRollback` will fail any invocation of `BytemanControlledXAResource.rollback()` until `greenFlag` is false (i.e. until we don't verify that XARecoveryModule sees the transaction to recover)

Reply:
Thank you for accommodating a change on the recovery manager - I will resolve this thread. Just to mention the second point is about the test currently named `testSuspensionWhenThereArentTransactionsToRecover`. I am I still resolving this thread though. Thank you